### PR TITLE
Add support for scopes oAuth apps

### PR DESCRIPTION
**Pull request by ivasilov**
This PR adds scopes to oAuth apps so that integrations can specify the needed permissions.
Screenshot 2023-10-06 at 12 18 57
Screenshot 2023-10-06 at 12 20 03

### CODE REVIEW

File: studio/components/interfaces/Organization/OAuthApps/PublishAppSidePanel/index.tsx [hunk 1]
```diff
@@ -139,6 +144,7 @@ const PublishAppSidePanel = ({
     return errors
   }
 
+  console.log(scopes)
```
Comment (alaister) [inline studio/components/interfaces/Organization/OAuthApps/PublishAppSidePanel/index.tsx:147]: Cheeky console.log

File: studio/components/interfaces/Organization/OAuthApps/AuthorizeRequesterDetails.tsx [hunk 2]
```diff
@@ -1,13 +1,68 @@
...
 }
 
-const AuthorizeRequesterDetails = ({ icon, name, domain }: AuthorizeRequesterDetailsProps) => {
+const AuthorizeRequesterDetails = ({
+  icon,
+  name,
+  domain,
+  scopes,
+}: AuthorizeRequesterDetailsProps) => {
+  console.log(scopes)
```
Comment (joshenlim) [inline studio/components/interfaces/Organization/OAuthApps/AuthorizeRequesterDetails.tsx:62]: one more here
```suggestion
```

File: studio/components/interfaces/Organization/OAuthApps/AuthorizeRequesterDetails.tsx [hunk 3]
```diff
@@ -1,13 +1,68 @@
...
+      .filter(Boolean)
+
+      .map((str) => (
+        <span key={str} className="font-semibold text-foreground">
+          {str}
+        </span>
+      ))
+      .reduce((acc, v) => (
+        <>
+          {acc}
```
Comment (joshenlim) [inline studio/components/interfaces/Organization/OAuthApps/AuthorizeRequesterDetails.tsx:36]: can we capitalize the actions here? Read and Write

File: studio/components/interfaces/Organization/OAuthApps/AuthorizeRequesterDetails.tsx [hunk 4]
```diff
@@ -18,13 +73,73 @@ const AuthorizeRequesterDetails = ({ icon, name, domain }: AuthorizeRequesterDet
...
+            allScopes={scopes}
+          />
+          <ScopeSection
+            description="access to all project's metadata, its upgrade status, network restrictions and network bans."
+            readScope={OAuthScope.PROJECTS_READ}
+            writeScope={OAuthScope.PROJECTS_WRITE}
+            allScopes={scopes}
+          />
+          <ScopeSection
+            description="access to all project's postgREST configurations."
```
Comment (joshenlim) [inline studio/components/interfaces/Organization/OAuthApps/AuthorizeRequesterDetails.tsx:126]:
```suggestion
            description="access to all project's PostgREST configurations."
```

File: studio/components/interfaces/Organization/OAuthApps/PublishAppSidePanel/Scopes.tsx [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        title="Organizations"
+        description="Organizations and all its members."
+        readScopeName={OAuthScope.ORGANIZATIONS_READ}
+        writeScopeName={OAuthScope.ORGANIZATIONS_WRITE}
+        scopes={scopes}
+        setScopes={setScopes}
+      />
+      <Scope
+        title="Projects"
+        description="Project's metadata, its upgrade status, network restrictions and network bans."
```
Comment (joshenlim) [inline studio/components/interfaces/Organization/OAuthApps/PublishAppSidePanel/Scopes.tsx:159]:
```suggestion
        description="Project's metadata, upgrade status, network restrictions, and network bans."
```

**Comment by joshenlim** [comment]
i think lets re-order the UI a little on the authorize screen, so instead of adding "Please note:... " at the bottom, shift it up top, and then we can also remove "all projects" in each of the perms, make it less wordy:

image

File: studio/components/interfaces/Organization/OAuthApps/AuthorizeRequesterDetails.tsx [hunk 6]
```diff
@@ -18,13 +73,73 @@ const AuthorizeRequesterDetails = ({ icon, name, domain }: AuthorizeRequesterDet
...
       </div>
+
       <p className="text-sm text-foreground-light">
-        {name} ({domain}) is requesting API access to an organization. The application will be able
-        to{' '}
-        <span className="text-amber-1200">
-          read and write the organization's settings and all of its projects.
-        </span>
+        Please note, the shown scopes will apply for the{' '}
+        <span className="text-amber-1200">selected organization and all of its projects.</span>
```
Comment (joshenlim) [inline studio/components/interfaces/Organization/OAuthApps/AuthorizeRequesterDetails.tsx:142]: lets also use text-amber-900 here to make it more visually constrasting

```suggestion
        <span className="text-amber-900">selected organization and all of its projects.</span>
```

File: studio/components/interfaces/Organization/OAuthApps/AuthorizeRequesterDetails.tsx [hunk 7]
```diff
@@ -1,13 +1,68 @@
...
   name: string
   domain: string
+  scopes: OAuthScope[]
+}
+
+const ScopeSection = ({
+  description,
+  readScope,
+  writeScope,
+  allScopes,
```
Comment (joshenlim) [inline studio/components/interfaces/Organization/OAuthApps/AuthorizeRequesterDetails.tsx:15]: just wondering - can we simplify this? we just need `hasRead` and `hasWrite` in this component right? should we just make those the arguments instead? then we don't need to pass `allScopes` as well - reduce the amount of props

`const ScopeSection = ({ description, hasRead, hasWrite }) => ...`

**Comment by joshenlim** [comment]
@ivasilov all looks good to me but can I trouble you to help check one more thing - how does the authorize page look like
cause i've got a feeling the modal is too long now for a macbook screen dimension (1440 x 980). It might be blocking the layout that the component sits in (there's a supabase logo top left corner iirc)

image

**Review by kangmingtay — COMMENTED** [review]
LGTM from the backend / API perspective - i've also tested it end-to-end and it looks good

Just a small feedback though, this can probably be addressed in a separate PR - can we address [this comment](https://github.com/supabase/supabase/blob/504a4ce4bfb7b52a6af30f2ef337c9eb4a04ab43/studio/components/interfaces/Organization/OAuthApps/OAuthApps.tsx#L18) and also show the approved scopes in the "Authorized Apps" section? Maybe we can show the approved scopes in a side panel?

**Review by joshenlim — APPROVED** [review]

**Pull request merged**